apps/emqx/src/emqx_tls_certfile_gc.erl

```diff
@@ -0,0 +1,335 @@
+    RootDir = emqx_utils_fs:canonicalize(emqx:mutable_certs_dir()),
+    Orphans = orphans(RootDir),
+    OrphansLast = maps:get(orphans, St, #{}),
+    Convicts = convicts(Orphans, OrphansLast),
+    Result = collect_files(Convicts, RootDir, Opts),
+    {ok, Result, St#{orphans => maps:without(Convicts, Orphans)}}.
+
+orphans() ->
+    Dir = emqx_utils_fs:canonicalize(emqx:mutable_certs_dir()),
+    orphans(Dir).
+
+orphans(Dir) ->
+    Certfiles = find_managed_files(fun is_managed_file/2, Dir),
+    lists:foldl(
+        fun(Root, Acc) ->
+            References = find_references(Root),
+            maps:without(References, Acc)
+        end,
+        Certfiles,
+        emqx_config:get_root_names()
+    ).
+
+convicts(Orphans, OrphansLast) ->
+    maps:fold(
+        fun(AbsPath, #file_info{mtime = MTime, inode = Inode}, Acc) ->
+            case maps:get(AbsPath, Orphans, undefined) of
+                #file_info{mtime = MTime, inode = Inode} ->
+                    % Certfile was not changed / recreated in the meantime
+                    [AbsPath | Acc];
+                _ ->
+                    % Certfile was changed / created / recreated in the meantime
+                    Acc
+            end
+        end,
+        [],
+        OrphansLast
+    ).
+
+find_managed_files(Filter, Dir) ->
+    emqx_utils_fs:traverse_dir(
+        fun
+            (AbsPath, Info = #file_info{}, Acc) ->
+                case Filter(AbsPath, Info) of
+                    true -> Acc#{AbsPath => Info};
+                    false -> Acc
+                end;
+            (AbsPath, {error, Reason}, Acc) ->
+                ?SLOG(notice, "filesystem_object_inaccessible", #{
+                    abspath => AbsPath,
+                    reason => Reason
+                }),
+                Acc
+        end,
+        #{},
+        Dir
+    ).
+
+is_managed_file(AbsPath, #file_info{type = Type, mode = Mode}) ->
+    % NOTE
+    % We consider a certfile is managed if: this is a regular file, owner has RW permission
+    % and the filename looks like a managed filename.
+    Type == regular andalso
+        ?HAS_OWNER_READ(Mode) andalso
+        ?HAS_OWNER_WRITE(Mode) andalso
+        emqx_tls_lib:is_managed_ssl_file(AbsPath).
+
+find_references(Root) ->
+    Config = emqx_config:get_raw([Root]),
+    fold_config(
+        fun(Stack, Value, Acc) ->
+            case is_file_reference(Stack) andalso is_string(Value) of
+                true ->
+                    Filename = emqx_schema:naive_env_interpolation(Value),
+                    {stop, [emqx_utils_fs:canonicalize(Filename) | Acc]};
+                false ->
+                    {cont, Acc}
+            end
+        end,
+        [],
+        Config
+    ).
+
+is_file_reference([<<"keyfile">> | _]) -> true;
+is_file_reference([<<"certfile">> | _]) -> true;
+is_file_reference([<<"cacertfile">> | _]) -> true;
+is_file_reference([<<"issuer_pem">>, <<"ocsp">> | _]) -> true;
+is_file_reference(_) -> false.
```

Comment:
How could we keep in sync this list with possible newly added certs?

Reply:
Good question. Speaking of this PR, it'd probably be a good idea to move that stuff to `emqx_tls_lib`. Generally speaking, my (optimistic) outlook is that we won't need to mess with this "reifying" process anymore, instead keep configured certs as is: either filename or certificate itself (i.e. if the user pasted it in the dashboard).